ocrd_network/ocrd_network/deployer.py

```diff
@@ -0,0 +1,302 @@
+from ocrd_network.deployment_config import *
+from ocrd_network.deployment_utils import (
+    create_docker_client,
+    create_ssh_client,
+    CustomDockerClient,
+    DeployType,
+    HostData,
+)
+from ocrd_network.rabbitmq_utils import RMQPublisher
+
+
+class Deployer:
+    """ Class to wrap the deployment-functionality of the OCR-D Processing-Servers
+
+    Deployer is the one acting. Config is for representation of the config-file only. DeployHost is
+    for managing information, not for actually doing things.
+    """
+
+    def __init__(self, config: ProcessingServerConfig) -> None:
+        """
+        Args:
+            config: Parsed processing-server-configuration
+        """
+        self.log = getLogger(__name__)
+        self.config = config
+        self.hosts = HostData.from_config(config.hosts)
+        self.mongo_pid = None
+        self.mq_pid = None
+
+    def kill_all(self) -> None:
+        """ kill all started services: workers, database, queue
+
+        The order of killing is important to optimize graceful shutdown in the future. If RabbitMQ
+        server is killed before killing Processing Workers, that may have bad outcome and leave
+        Processing Workers in an unpredictable state
+        """
+        self.kill_hosts()
+        self.kill_mongodb()
+        self.kill_rabbitmq()
+
+    def deploy_hosts(self, rabbitmq_url: str, mongodb_url: str) -> None:
+        for host in self.hosts:
+            self.log.debug(f'Deploying processing workers on host: {host.config.address}')
+
+            if (any(p.deploy_type == DeployType.native for p in host.config.processors)
+                    and not host.ssh_client):
+                host.ssh_client = create_ssh_client(
+                    host.config.address,
+                    host.config.username,
+                    host.config.password,
+                    host.config.keypath
+                )
+            if (any(p.deploy_type == DeployType.docker for p in host.config.processors)
+                    and not host.docker_client):
+                host.docker_client = create_docker_client(
+                    host.config.address,
+                    host.config.username,
+                    host.config.password,
+                    host.config.keypath
+                )
+
+            for processor in host.config.processors:
+                self._deploy_processing_worker(processor, host, rabbitmq_url, mongodb_url)
+
+            if host.ssh_client:
+                host.ssh_client.close()
+            if host.docker_client:
+                host.docker_client.close()
+
+    def _deploy_processing_worker(self, processor: WorkerConfig, host: HostData,
+                                  rabbitmq_url: str, mongodb_url: str) -> None:
+
+        self.log.debug(f'deploy \'{processor.deploy_type}\' processor: \'{processor}\' on'
+                       f'\'{host.config.address}\'')
+
+        for _ in range(processor.count):
```

Comment:
What about failover? IIUC if a processor crashes, ATM that worker slot will just be gone.

Reply:
Right. There isn't a mechanism to track currently active ("live") processing workers so they can be started again when crashed. We need better worker/process management. For now, we kept such details off.

Reply:
I know the topic is much bigger, but in the simplest form, it could be a simple while-catch loop (and logging the exception of course). Next best would be a CLI option controlling this, say `--worker-failure [raise|ignore|restart]` (where `restart` would enter the loop, `ignore` would let the worker die, `raise` would terminate the whole server.

It's very relevant – we have crashes frequently, and processors themselves have no error handling ATM.

Reply:
> We should differentiate between worker crash and processor crash though and not use both terms interchangeably. We have exception logging for the processor crash. We also have logging for the worker crash as well. We just lack mechanism to restart the crashed worker. Crashed processor does not result into a crashed worker.

Oh right! So this would cover the queue and db handling parts.

> If a processor crashes due to an external error (e.g. workspace related issue) the processor fails for that specific processing message execution but the worker itself does not fail. The worker still marks that the message was processed successfully with status "stopped" i.e. job failed and proceeds by getting the next processing message from the queue.

Yes, and so a higher-level component (say workflow server) could act on the failed job by re-scheduling etc. right?

BTW, for a discussion of lower-level error handling (in the processor itself, i.e. at page level) see [here](https://github.com/OCR-D/core/issues/579). But re-entry of partially run workflows is a cross-over topic.

Reply:
> Yes, and so a higher-level component (say workflow server) could act on the failed job by re-scheduling etc. right?

Fault tolerance and error handling topics are complex. Depends. From a higher-level component perspective, yes - but only if that higher-level component knows how to fix the problem and resubmit the job. Take a look at the fault tolerance and idempotence description (coordination project task in the ZenHub board) of the processing worker:

> **Fault tolerance**
To increase fault tolerance, manual acknowledgement mode must be used. In this case,
    if the processing finished successfully, send a positive ACK signal.
    if the processing ended in failure, try again 3 times. If it's still failed, send a negative ACK signal to let the queuing system re-queue the message.
If there is no ACK received (e.g., consumer crash), the queuing system will automatically re-queue the message after timeout. [By default](https://www.rabbitmq.com/consumers.html#acknowledgement-timeout), it's 30 minutes. This value might need to be adjusted to fit best to our use case.
**Idempotence**
If there is something wrong, e.g. network, the message will be sent again to the consumer. Make sure to check the [redeliver](https://www.rabbitmq.com/confirms.html#automatic-requeueing) property to make sure that we don't process a page twice.

According to that description:
1. The processing worker restarts the OCR-D processor with the same parameters 3 times (**Note**: To keep it simple the current implementation runs only once and fails once, not 4 times - is there even a case when we would like to run more than once here? Assuming that failing once will probably also fail 4 times if nothing in the processing message is changed.)
2. The processing worker sending negative ACK (NACK) does not really make sense here if the failed processing message is requeued as it is again without any changes. A failed message will still fail. The only time when it makes sense to send a NACK is when say the processing worker receives a termination signal and catches it (not supported in the current state) - then the processing message can be requeued back. In that case, it is still unknown whether the processing message will potentially fail or not in the future due to external errors, e.g., in the workspace. (**Note**: in either case, the workspace will be modified to a certain extent and we need a way to track this - a backup mechanism or extension of the Mets caching in the core. Of course, with page-level processing this becomes an easier task.)
3. Ideally, the failed processing message should be forwarded to another queue ([dead letter exchange](https://www.rabbitmq.com/dlx.html)) to avoid an infinite loop of requeuing failing messages to the same queue.
4. The redeliver flag/property of the processing message hints that the message was requeued so either of these happened:
- The message was NACKed due to consumer timeout (worker has failed or still processing)
- The message was NACKed manually (the processor has failed 4 times or the worker was terminated and caught the termination signal to NACK the message before termination). 

However, simply by checking that property, we could not really say whether: 
- The message was already processed - don't process it again/twice (as described above)
- The message was not (fully) processed - (re)process it. In that case, we again need to know exactly where to continue from. In case the processing message is a single page, then I guess it's not a big deal to process the entire page again with the `--overwrite` flag.

5. The Workflow Server can check the processing job status in the DB / Result queue. If the processing job has failed, the Workflow Server knows exactly which job failed and can resubmit the job through the Processing Server. Of course, assuming that the Workflow Server is smart enough to be able to fix the error (which is again a broad topic). Potentially, user involvement will be required here to manually submit the job after fixing the external error.

More regarding the workflow execution:
Lately, I have been busy with implementing a Nextflow script to be executed by the Workflow Server that does the following:
1 Sends a processing job to the Processing Server
2 Connects a RabbitMQ consumer to listen to the result queue of the processor specified in the job
3 After consuming a single message the consumer is disconnected 
4 Based on the job state: 
   - Success - the next OCR-D process in the workflow is executed
   - Fail - the workflow is terminated

However, still facing a problem that I was not able to solve yet - synchronization between the blocking consumer and the parent process that executes the Nextflow processes inside the Nextflow workflow. The thing is that I have been also digging in the Nextflow source code to get a better understanding of how processes and threads are utilized since in the documentation itself it's not obvious what is a separate process and what is a separate thread of a specific thread (and honestly, why would they clarify such detail when not relevant for a big portion of the users). What I am planning is to try a polling strategy. Since the workflow cannot continue without having the status of the previous job, this still makes sense.

I am not sure if that would be the elegant approach to running workflows - i.e. with a RabbitMQ consumer inside the Nextflow script. Especially, since the Nextflow scripts have to be submitted by the users. On the other hand, if the Workflow Server creates worker processes for submitting processing jobs and consumers for consuming from result queues - we lose the point of having the Nextflow in the middle between the Workflow Server and the Processing Server. 

Hope this helps to get more insights.

> BTW, for a discussion of lower-level error handling (in the processor itself, i.e. at page level) see https://github.com/OCR-D/core/issues/579. But re-entry of partially run workflows is a cross-over topic.

Thanks.

Reply:
> According to that description:

Interesting, thanks for sharing here!

> 1. The processing worker restarts the OCR-D processor with the same parameters 3 times

_re-runs_ the processor, yes

> (**Note**: To keep it simple the current implementation runs only once and fails once, not 4 times - is there even a case when we would like to run more than once here? Assuming that failing once will probably also fail 4 times if nothing in the processing message is changed.)

Don't be too sure about that. In mass production context, we will likely see temporary disk outages, OOM due to oversubscription (both on CPU and GPU), race conditions due to bad processor code etc.

But I would suggest making this configurable for the admin, so one can trade robustness for efficiency. (The operators will know whether they prefer to spend CPU time to get as much done as possible or to ensure high coverage of the materials attempted.)

Also, again, the processor itself must have some sensible error handling, page-wise.

Reply:
> 2\. The processing worker sending negative ACK (NACK) does not really make sense here if the failed processing message is requeued as it is again without any changes. A failed message will still fail.

Same as above. It _could_ still work another time. Also, another worker instance might have more luck. (Could be equipped with better runtime resources, or installed with newer software versions etc.)

> The only time when it makes sense to send a NACK is when say the processing worker receives a termination signal and catches it (not supported in the current state) - then the processing message can be requeued back. 

Yes, that makes sense.

> In that case, it is still unknown whether the processing message will potentially fail or not in the future due to external errors, e.g., in the workspace.

True. If the workspace broke or became unavailable on disk, then no further attempts make sense. So the error could even be **escalated** (up to the workflow server if necessary).

> (**Note**: in either case, the workspace will be modified to a certain extent and we need a way to track this - a backup mechanism or extension of the Mets caching in the core. Of course, with page-level processing this becomes an easier task.)

I don't think we risk inconsistency ATM. If the processor fails, the METS will stay untouched. If the worker fails before the processor succeeds, same thing. If the worker fails after the processor succeeds, the worst thing that can happen is that we repeat the work – but for that we will first need to enable _overwrite mode_.

For page-level processing, it gets more complicated IMO. We'll have more frequent METS updates, and lots of concurrent jobs on the same workspace. Worker failure for a single page should behave like processor failure for a single page (i.e. retry and fall back), but this must be reflected here, and differs from the failure strategies for the (single step on the) whole document.

Reply:
> 3\. Ideally, the failed processing message should be forwarded to another queue ([dead letter exchange](https://www.rabbitmq.com/dlx.html)) to avoid an infinite loop of requeuing failing messages to the same queue.

Couldn't we track some hop attribute on the messages to break out of such loops eventually?

I don't understand – how does the DLX work?

Reply:
> 4\. The redeliver flag/property of the processing message hints that the message was requeued so either of these happened:
> 
>   * The message was NACKed due to consumer timeout (worker has failed or still processing)
>   * The message was NACKed manually (the processor has failed 4 times or the worker was terminated and caught the termination signal to NACK the message before termination).
> 
>   However, simply by checking that property, we could not really say whether:
> 
>   * The message was already processed - don't process it again/twice (as described above)

So either add some hop count attribute (and use a threshold when to cancel the job entirely), or simply prevent rescheduling messages which already have redeliver flag set (which would amount to a hop threshold of 1).

>   * The message was not (fully) processed - (re)process it. In that case, we again need to know exactly where to continue from. In case the processing message is a single page, then I guess it's not a big deal to process the entire page again with the `--overwrite` flag.

Apart from the single-page processing scenario (which I don't want to get into here), I don't see this possibility. The only case I could think of is a worker failure following a processor success (see above).

Reply:
> 5\. The Workflow Server can check the processing job status in the DB / Result queue. If the processing job has failed, the Workflow Server knows exactly which job failed and can resubmit the job through the Processing Server. Of course, assuming that the Workflow Server is smart enough to be able to fix the error (which is again a broad topic). Potentially, user involvement will be required here to manually submit the job after fixing the external error.

Much of the above speaks to the necessity to (somehow) mark the **reason** for failures at the message level, discerning between temporary outtime and unfixable problems.

There is also another case – at the workflow / API level: [evaluator](https://github.com/OCR-D/spec/issues/172) failures. The latter would allow triggering alternative workflow routes. Naturally, that distinction (fail vs. flop) must even be coded at the user level. So if the overall workflow has no alternative routes anymore, the user should be notified to take action (like browsing the interim results and reconfiguring the workflow).

But for the purely technical failures I don't see the user in the loop. If there is a technical problem, the admin should take action.

Reply:
> More regarding the workflow execution:
> Lately, I have been busy with implementing a Nextflow script to be executed by the Workflow Server that does the following: [...]

I've seen it (somewhere), but was surprised that you would want to solve this via Nextflow. In my understanding, the natural implementation of the Workflow Server would be via FastAPI Python code as well.

Reply:
> I am not sure if that would be the elegant approach to running workflows - i.e. with a RabbitMQ consumer inside the Nextflow script. Especially, since the Nextflow scripts have to be submitted by the users.

Yes, that makes it worse IMO. All this code clutters the actual workflow description. The NF workflow format should be as simple and declarative as possible. (Workflow configurations will be managed over time, most likely evolving from each other by publication and customisation. And it must be easy for users to understand and adapt a configuration. All this would be severely hampered if they were to contain code which actually belongs to the WF server logic.)

> On the other hand, if the Workflow Server creates worker processes for submitting processing jobs and consumers for consuming from result queues - we lose the point of having the Nextflow in the middle between the Workflow Server and the Processing Server.

Yes. The NF script should more or less look like a controlled shell syntax (processor and evaluator calls, possibly intertwined with control structures). It should be up to the Workflow Server's implementation to use CLI calls or Processing Server calls (with result polling or result queue listening). Perhaps we should rethink this as a custom _executor_?

Reply:
> > It _could_ still work another time.  Also, another worker instance might have more luck. (Could be equipped with better runtime resources, or installed with newer software versions etc.)
> 
> There is no guarantee that another worker instance will consume the failed message.

True, but there's also a chance the original worker succeeds on reconsumption – it depends on the cause. Of course you can argue against repeated consumption in general / on average (low likelihood, waste of resources). But again, let the admin decide!

> The easiest solution for our use case would be to add a hop counter to OcrdProcessingMessage.

Yes.

> > I don't think we risk inconsistency ATM. If the processor fails, the METS will stay untouched.
> 
> Is that always the case or is it OCR-D processor implementation-dependent (i.e. when the update workspace method is invoked)?

Granted, it's not enforced, just a convention: So far all processors loop over pages and do METS updates in memory and PAGE output on filesystem, before the final write-back (which could become a METS server call soon).

With the planned processor API changes, we will have the freedom to decide when to update the METS centrally (via `ocrd.Processor.process_workspace` which gets inherited and does the looping). In particular, we could provide a synchronisation mechanism for page-wise processing across workflow steps in that new implementation. (For example, multi-threading across pages and blocking on missing input pages.) But if we implement page-wise processing in the Processing/Workflow Server – like an "array" of normal jobs for single pages –, then indeed the interdependency might be lost. (So at least the Workflow Server must know "page 3 in step 2 depends on page 3 in step 1 only".) Either way, single-page failure would not mean rollback or influence other pages, only require skip or fallback. (So the planned single-page error handling will have to be lifted up outside of the processor's loop.)

> > Much of the above speaks to the necessity to (somehow) mark the reason for failures at the message level, discerning between temporary outtime and unfixable problems.
> 
> My initial design of the OcrdResultMessage contained both a `status` and a `return` field. I guess the significance of that `return` field was not relevant then and was dropped in the feedback. The idea behind the `return` field was to store the return code of the processor or any code set by the processing worker when the processing fails.

We don't have a spec on processor return codes though. With a suitable exception ladder, we don't need that IMO. If a page failed on the processor, we can have skip, fallback, and raise. The latter could differentiate between unfixable (bug on data path, or max-hop) and temporary (OOM, I/O, timeout). So on the queue, it would suffice to differentiate between job status transient failure and persistent failure, and hop count in the requeued message (or the job object).

BTW, also confer Nextflow's own error handling via [errorStrategy](https://www.nextflow.io/docs/latest/process.html#errorstrategy) and [maxErrors](https://www.nextflow.io/docs/latest/process.html#maxerrors)...

(But again, the additional distinction failure vs flop must also be addressed.)


> The Workflow endpoint is in fact implemented via FastAPI Python code. The Workflow Server currently triggers a Nextflow Process which executes the specified Nextflow script (uploaded earlier by the user or admin). Inside that script, the workflow logic (i.e. chaining of processors) is described. The workflow server is supposed to just manage Nextflow processes - start them and restart/resume them when failed. Thus, there is a multi-level of error handling which should be but is still not defined precisely.

Okay, now I understood, thanks!

> Check this working Nextflow script [example](https://github.com/MehmedGIT/wf_server_nf_script/blob/master/wf_server_example.nf). Currently, everything is in a single file. Ideally, the user should implement only the `processor` and `workflow` blocks (starting from line 150). High-level methods such as `exec_block_logic()` should be provided to the user as a library. The user should then only care about the call parameters of the OCR-D processor and potentially some workflow parameters. Parameters related to the RabbitMQ will be injected by the Workflow Server.

Fascinating! This opens some [questions](https://github.com/MehmedGIT/wf_server_nf_script/issues/1) right away for me.

Reply:
> True, but there's also a chance the original worker succeeds on reconsumption – it depends on the cause. Of course you can argue against repeated consumption in general / on average (low likelihood, waste of resources). But again, let the admin decide!

Honestly, I don't have a strong opinion on that. I agree that it's the best to make it configurable and let the admin decide.

> (So at least the Workflow Server must know "page 3 in step 2 depends on page 3 in step 1 only".) Either way, single-page failure would not mean rollback or influence other pages, only require skip or fallback. (So the planned single-page error handling will have to be lifted up outside of the processor's loop.)

That's how I understand it as well. This also how all previous pseudo page-wise implementations in Nextflow were done (with splitting and merging the mets files). There is no dependency between the pages. Just a dependency between the steps of each page. A failed page just breaks its follow-up steps. 
Note that in case multiple pages (pages range) is passed to a single step (within a Nextflow process block), then a single broken page will also break the process block of that step and the follow up steps of all pages in the range.

> We don't have a spec on processor return codes though.

We don't really need a spec on processor return codes. The processing worker may assign the return codes based on the caught exception message from the failed processor (although I don't like that approach).

> So on the queue, it would suffice to differentiate between job status transient failure and persistent failure, and hop count in the requeued message (or the job object).

Maybe splitting the FAILURE status to two, say: "TRANSIENT FAILURE" and "PERSISTENT FAILURE" as a job status may be enough then? 

> BTW, also confer Nextflow's own error handling via [errorStrategy](https://www.nextflow.io/docs/latest/process.html#errorstrategy) and [maxErrors](https://www.nextflow.io/docs/latest/process.html#maxerrors)...

I am aware of these process directives. However, these are not really usable in the context of processing workers and OCR-D processors. The `errorStrategy` and `maxErrors` will apply to the process blocks which just send processing messages to the processing server and reading the results from the result queue or DB (the Nextflow script exampel I provided above).

These 2 directives are really useful when the process block contains ocr-d processor calls. Consider this simple example below. The `foo` process is executed in the HPC and if the task fails due to out of memory error, the next execution attempts will increase the memory allocation and try again. Also note that the task exit status codes are not standard and can change depending on the resource manager used in the HPC cluster.

```
process binarization {
  memory { 2.GB * task.attempt }
  errorStrategy { task.exitStatus in 137..140 ? 'retry' : 'terminate' }
  maxRetries 3

  script:
  """
  call the ocr-d processor here
  """
}
``` 
Of course, this may also be useful for Workflow Server level error handling.

> Fascinating! This opens some https://github.com/MehmedGIT/wf_server_nf_script/issues/1 right away for me.

Thanks, feedback and questions are always appreciated.

Reply:
> That's how I understand it as well. This also how all previous pseudo page-wise implementations in Nextflow were done (with splitting and merging the mets files). 

… which will become easier and more reliable with the METS server (no more splitting/merging, not even wasting I/O bandwidth for METS de/serialisation between workflow steps – only after final step). If done right.

> There is no dependency between the pages. Just a dependency between the steps of each page. A failed page just breaks its follow-up steps.

Yes, the dependency can go away when the METS is not a shared resource anymore, but a server.

> Note that in case multiple pages (pages range) is passed to a single step (within a Nextflow process block), then a single broken page will also break the process block of that step and the follow up steps of all pages in the range.

Good point. So page ranges beyond single page should have a different failure semantic: We cannot simply skip or fallback, because the other affected pages might be "innocent", therefore persistent failures must escalate.

> We don't really need a spec on processor return codes. The processing worker may assign the return codes based on the caught exception message from the failed processor (although I don't like that approach).

Yes, that's what I meant by exception ladder. (Sorry, again had confused processor result with worker result.)

> Maybe splitting the FAILURE status to two, say: "TRANSIENT FAILURE" and "PERSISTENT FAILURE" as a job status may be enough then?

Is what I'm sayin, yes. (It can only be guesswork, though. But persistent after max-tries is at least better than unconditionally persistent. And we can still have the admin set max-tries to 1.)

> > BTW, also confer Nextflow's own error handling via [errorStrategy](https://www.nextflow.io/docs/latest/process.html#errorstrategy) and [maxErrors](https://www.nextflow.io/docs/latest/process.html#maxerrors)...
> 
> I am aware of these process directives. However, these are not really usable in the context of processing workers and OCR-D processors.

Yes – I was already back at the Workflow Server via Nextflow level.

> These 2 directives are really useful when the process block contains ocr-d processor calls. Consider this simple example below. The `foo` process is executed in the HPC and if the task fails due to out of memory error, the next execution attempts will increase the memory allocation and try again. Also note that the task exit status codes are not standard and can change depending on the resource manager used in the HPC cluster.
> 
> ```
> process binarization {
>   memory { 2.GB * task.attempt }
>   errorStrategy { task.exitStatus in 137..140 ? 'retry' : 'terminate' }
>   maxRetries 3
> ```

Nice! 

Does the SLURM executor acknowledge Nextflow's runtime resources universally (across HPC installations) and exhaustively (including things like GPU and number of cores)?

Reply:
> So page ranges beyond single page should have a different failure semantic: We cannot simply skip or fallback, because the other affected pages might be "innocent", therefore persistent failures must escalate.

Exactly.

> ... the SLURM executor acknowledge Nextflow's runtime resources ...

What do you mean by that? Initially, the user allocates resources (memory and CPU) from the SLURM scheduler for the Nextflow workflow job. Then once the workflow job starts executing, the Nextflow itself submits new processing requests to the SLURM scheduler. Each Nextflow process can specify separate resource needs as a process block directive or through the `nextflow.config` file. The upper limit of the resources for each Nextflow process is the number of resources requested for the entire Nextflow workflow job. The Nextflow itself creates an execution report with the CPU and memory consumption of each Nextflow process. I didn't see any GPU-related metrics anywhere in the Nextflow report or documentation. 

The user needs to know beforehand what kind of GPUs are available inside the HPC to be able to reference how many GPUs and which type should be requested individually for each process with the `accelerator` directive. E.g.:

```nextflow
process some_gpu_heavy_processor {
    accelerator 6, type: 'nvidia-gtx-1080'

    script:
    """
    ocrd processor call
    """
}
```
will allocate 6 GPU cores of the specified type.

Reply:
> My question was whether these are exhaustively accessible from NF, without being dependent on the concrete NF and HPC installation (e.g. name of GPU).

I am still not sure whether I understand entirely what `exhaustively accessible` refers to.

Maybe a simple example could help (although even a simple example is not that simple). 
Consider the following 3 files and their contents:
<details>
<summary> batch_script.sh </summary>

```bash
#!/bin/bash
#SBATCH --partition medium
#SBATCH --constraint scratch
#SBATCH --cpus-per-task 4
#SBATCH --mem 16G
#SBATCH --output /home/users/mmustaf/jobs_output/job-%J.txt

# clear the environment then load singularity and nextflow
module purge
module load singularity  # loads "git" and "go" as well
module load nextflow  # loads "openjdk" as well

... 
# managing the creation of directories and moving files here 
...

# execute the main Nextflow script
nextflow run ./simple.nf --file_group "DEFAULT" --volumedir "path/to/workspace"

... 
# transferring the results from the scratch storage to somewhere else
# cleaning no longer needed files
...
```
</details>

<details>
<summary> nextflow.config </summary>

```nextflow
singularity {
  enabled = true
  // Path to the singularity SIF file of the OCR-D all software
  cacheDir = '/scratch1/users/mmustaf/singularityCache'
}

executor {
  name = 'slurm'
  // ...
  // Other SLURM related configurations 
  // ...
}

process {
  withName: download_workspace {
    cpus = 1
    memory = 8.GB
    queue = 'medium'  // partition
  }
  withName: ocrd_cis_ocropy_binarize {
    cpus = 4
    memory = 16.GB
  }
  withLabel: active_gpu {
     queue = 'gpu'  // partition
     containerOptions = '--nv'  // to make GPU drivers accessible in the singularity container
  }
}
```
</details>

<details>
<summary> simple.nf </summary>

```nextflow
nextflow.enable.dsl=2

params.file_group = ""
params.volume_dir = ""
// $projectDir is the directory where this NF script is located
params.mets_path = "$projectDir/ocrd-workspace/mets.xml"

process download_workspace {
  input:
    val file_group
  output:
    val file_group

  script:
  """
  singularity exec --bind ${params.volume_dir} docker://ocrd/all:maximum ocrd workspace find --file-grp ${file_group} --download --wait 1
  """
}

process ocrd_cis_ocropy_binarize {
  label 'active_gpu'
  input:
    path mets_file 
    path dir_name
  output:
    val "OCR-D-BIN"
  
  script:
  """
  singularity exec --bind ${params.volume_dir} docker://ocrd/all:maximum ocrd-cis-ocropy-binarize -m ${mets_file} -I ${dir_name} -O "OCR-D-BIN"
  """
}

// This is the main workflow
workflow {
  main:
    download_workspace(params.file_group)
    ocrd_cis_ocropy_binarize(params.mets_path, download_workspace.out)
}
```
</details>

NOTE: The examples are not tested/executed, and are provided as a reference.

1. The `batch_script.sh` is submitted to the SLURM scheduler through the front-end node of the HPC. The SLURM scheduler then starts the workflow job, i.e., the submitted Nextflow script with the amount of resources and conditions specified inside the batch script.
2. The SLURM executor of Nextflow then manages the allocation of separate jobs for each Nextflow process by talking to the SLURM scheduler of HPC and requesting resources specified inside the `nextflow.config` file. The upper limit for a resource is the resource allocated when executing the Nextflow script. In this example, 4 CPUs and 16GBs of memory.
3. With the `nextflow.config` (check [here](https://www.nextflow.io/docs/latest/config.html#config-scopes) regarding the scopes) it's possible to configure resources to be allocated for each step inside the NF script. So, that single configuration file could be used with any number of NF scripts as long as 1) the process names are consistent between the two files, or 2) specific labels are added inside the NF script inside processes as a directive to tag them into a group.
4. Using the GPU resource is possible when running the job inside the GPU partition (according to the [GWDG documentation](https://docs.gwdg.de/doku.php?id=en:services:application_services:high_performance_computing:running_jobs_slurm), I don't know about other HPC environments). So inside the batch script the partition parameter should be replaced and a the GPU unit specified with format `name:cores`:
```bash
#SBATCH --partition gpu
#SBATCH -G gtx1080:6
```
or just allocating 6 cores of any available GPU: `#SBATCH -G 6`. 

To be honest, I never used a GPU when running a workflow inside the HPC. So, I am not sure whether problems may arise from the fact that I am requesting the `medium` partition inside the `batch_script.sh` but then trying to execute the binarization processor inside the `gpu` partition. I can tell more once I play with different configurations. 

Overall, the hard part is to estimate the required resources for each step based on the provided workspace. To provide more flexible configurations, probably it would be required to produce the batch scripts and the nextflow config files dynamically. 

I think this post is generally more relevant for Operandi, but it still provides (hopefully) useful insights about the HPC environment, Nextflow, and OCR-D in general.

Reply:
Thanks for the illustration. I recommend documenting somewhere more lasting and accessible.

> 2\. The upper limit for a resource is the resource allocated when executing the Nextflow script. In this example, 4 CPUs and 16GBs of memory.

Ok, that answers the gist of my question. We can only sub-allocate resources here. Nextflow's Slurm executor does not get to create new Slurm jobs as first-class citizens. It _is_ a controlled vocabulary, but the chosen values get passed on directly to Slurm.

> So, I am not sure whether problems may arise from the fact that I am requesting the `medium` partition inside the `batch_script.sh` but then trying to execute the binarization processor inside the `gpu` partition.

Based on your description, I doubt it will work. (Slurm knows job dependencies and can thus chain jobs on different partitions, but you cannot "break out" of a partition AFAIK.)

> Overall, the hard part is to estimate the required resources for each step based on the provided workspace. 

I know. Very hard. Set it too low, and it will crash and you'll have to restart. Set it too high and the job will take much longer to be scheduled, and likely waste time (which cannot be otherwise filled by the scheduler). And esp. memory depends on the data, even various from page to page.

That's why I liked your NF repeat escalator so much. But it does not help us here – we'd have to determine the overall resources in the script beforehand anyway.

> To provide more flexible configurations, probably it would be required to produce the batch scripts and the nextflow config files dynamically.

Yes, that I was hoping the NF Slurm executor would do.

> I think this post is generally more relevant for Operandi, but it still provides (hopefully) useful insights about the HPC environment, Nextflow, and OCR-D in general.

We came here by discussing NF's own means of error handling. It's still the right place IMO, as it's about the overall architecture.

Reply:
> Thanks for the illustration. I recommend documenting somewhere more lasting and accessible.

I will edit it a bit and append it at the end of the Operandi repo readme file (for now). 

> We can only sub-allocate resources here. Nextflow's Slurm executor does not get to create new Slurm jobs as first-class citizens.

Yes. I would say it creates new slave slurm jobs for each process but they are restricted to the allocated resources for the entire workflow job.

> (Slurm knows job dependencies and can thus chain jobs on different partitions, but you cannot "break out" of a partition AFAIK.)

Well, the process does not really break out of a partition, but starts in a different partition than the one specified in the batch script.

> Yes, that I was hoping the NF Slurm executor would do.

How would Slurm executor know itself how much to allocate unless there is an internal module that monitors previous runs and learns to estimate based on previous runs. Sadly, no such magic is available.

Reply:
> > Yes, that I was hoping the NF Slurm executor would do.
> 
> How would Slurm executor know itself how much to allocate unless there is an internal module that monitors previous runs and learns to estimate based on previous runs. Sadly, no such magic is available.

Nothing in Slurm looks like it can learn anything. What I was expecting was merely that the Slurm executor allocates resources from the NF config. That's why I was asking regarding the relationship between both perspectives (Slurm resource specs and NF resource specs).

Reply:
@bertsky, is anything left here to be resolved before the merge? The error handling (e.g. restarting a failed processing worker) will come later with all the error handling implementation. I will transfer the valuable information under this conversation [here](https://pad.gwdg.de/rktJKPO8SGW4tCd1rwTZgQ?both). This pad should also serve as a guide what are the error handling requirements and what we have covered so far.

Reply:
Let's see:
> - the proper use of logging (Mehmed said he's waiting for Konstantin's input)

still unresolved.

> - the right place for the getcwd to save the old state in run_processor – Mehmed reverted the change because it fails some test he says

still [pending](https://github.com/OCR-D/core/pull/974/files#r1134717424).

The second one is a blocker IMHO.

Reply:
Yes, these 2 are a must before the merge. I wonder if there is anything left to be resolved from this specific conversation regarding the error handling.

Reply:
Oh, got it!

I'll work through the pad and let you know so we can resolve here.

Reply:
> Let's see:
> 
> > * the proper use of logging (Mehmed said he's waiting for Konstantin's input)
> 
> still unresolved.
> 
> > * the right place for the getcwd to save the old state in run_processor – Mehmed reverted the change because it fails some test he says
> 
> still [pending](https://github.com/OCR-D/core/pull/974/files#r1134717424).
> 
> The second one is a blocker IMHO.

Both aspects are my responsibility and I will fix them before release (though not necessarily before merge).

Reply:
@kba, just the logging is left for now. The second was fixed with #987 but there's #1017 as a result.

Reply:
@kba @MehmedGIT  What about the actual worker restart mechanism orginally addressed by this thread (before we had a lift-off on the lower/processor-level and higher/workflow-level error handling)?

> > I know the topic is much bigger, but in the simplest form, it could be a simple while-catch loop (and logging the exception of course). Next best would be a CLI option controlling this, say `--worker-failure [raise|ignore|restart]` (where `restart` would enter the loop, `ignore` would let the worker die, `raise` would terminate the whole server.
> 
> Agree and like the CLI option.
> 
> We should differentiate between worker crash and processor crash, though, and not use both terms interchangeably. We have exception logging for the processor crash. We also have logging for the worker crash as well. We just lack mechanism to restart the crashed worker. (Crashed processor does not result into a crashed worker.)

Reply:
> @bertsky, is anything left here to be resolved before the merge? The error handling (e.g. restarting a failed processing worker) will come later with all the error handling implementation.

By later, I meant after this PR is merged. Although, I think implementing this should be fairly easy and can be done before the merge in the `processing_worker_cli`.

The #1015 still has not got enough attention.

Reply:
Ok, sorry again for my bad memory. I think we agreed it would be okay to implement this later, along with error handling on the other levels. (Because the workers themselves are not so likely to fail anyway, and because it still must be properly covered what to do with messages that the failing worker takes to its grave etc.)

Resolved it is then!